File: gui_apps/ch01/first/first.cc
// -------------------------------------------------------------------------
// first.cc: Hello World with Qt
// this application creates a small window & displays a welcome message
// to the user along with the version of the Qt Framework in use.
//
// @author: Manish Bhobe
// My experiments with C++ & Qt Framework
// This code is meant for learning & educational purposes only!!
// -------------------------------------------------------------------------
#include <QApplication>
#include <QDesktopWidget>
#include <QLabel>
#include <QString>
#include <QTextStream>
#include <QVBoxLayout>

void showCenteredOnDesktop(QWidget &window)
{
   // center window on desktop
   // @see: http://www.qtcentre.org/threads/43158-How-to-center-main-window-on-screen
   window.show();
   QRect screenSize = QApplication::desktop()->screenGeometry();
   int x = (screenSize.width() - window.width()) / 2;
   int y = (screenSize.height() - window.height()) / 2;
   window.move(x, y);
}

int main(int argc, char **argv)
{
   QApplication app(argc, argv);

   // create & show the GUI
   QWidget window;
   window.setWindowTitle("Hello Qt World!");
   QString ver;
   QTextStream out(&ver);
   // is actually populating the underlying QString
   out << "You are using Qt " << QT_VERSION_STR;
   QLabel *version = new QLabel(ver);
   QVBoxLayout *l = new QVBoxLayout;
   l->addWidget(new QLabel("Hello World! Welcome to the Qt Framework."));
   l->addWidget(version);
   window.setLayout(l);

   // window.show();
   showCenteredOnDesktop(window);

   return app.exec();
}

File: gui_apps/ch01/first/first.py
# first.py : Hello World with PyQt5
import sys
import pathlib
from PyQt6.QtCore import *
from PyQt6.QtGui import *
from PyQt6.QtWidgets import *

USING_PYQT6 = True if PYQT_VERSION_STR.startswith('6') else False

p = pathlib.Path(__file__)
print(f"Current file's path is {str(p)}")
p.parents[2]
print(f"p.parents[2] = {str(p.parents[2])}")
sys.path.append(str(p.parents[2]))
import globalvars
globalvars.USING_PYQT6 = USING_PYQT6
import utils


class First(utils.CenteredOnDesktopWidget):
    def __init__(self, *args, **kwargs):
        super(utils.CenteredOnDesktopWidget, self).__init__(*args, **kwargs)
        msg = f"This is a PyQt{PYQT_VERSION_STR[0]} application!\n"
        msg = msg + f"You are using Qt version {QT_VERSION_STR}"
        self.label = QLabel(msg)
        self.setWindowTitle("Hello PyQt")
        layout = QVBoxLayout()
        layout.addWidget(self.label)
        self.setLayout(layout)


def main():
    app = QApplication(sys.argv)
    app.setFont(QApplication.font("QMenu"))

    w = First()
    w.show()
    return app.exec()


if __name__ == '__main__':
    main()
